stroyprombeton/migrations/0019_catalog_data.py

```diff
@@ -0,0 +1,127 @@
+    'length': 'Длина',
+    'width': 'Ширина',
+    'height': 'Высота',
+    'weight': 'Масса',
+    'volume': 'Объём',
+    'diameter_out': 'Внешний диаметр',
+    'diameter_in': 'Внутренний диаметр',
+}
+
+TAG_FIELDS = FIELD_TO_TAG_GROUP.keys()
+
+OPTION_FIELDS = [
+    'date_price_updated', 'code', 'mark', 'is_new_price', 'price', 'in_stock',
+]
+
+
+class CreateOptionsOperation(Operation):
+    def state_forwards(self, app_label, state):
+        pass
+
+    @transaction.atomic()
+    def database_forwards(self, app_label, schema_editor, _, to_state):
+        product_model = to_state.apps.get_model(app_label, 'Product')
+        option_model = to_state.apps.get_model(app_label, 'Option')
+
+        grouped_products = groupby(
+            product_model.objects.all(),
+            key=attrgetter('name')
+        )
+        for name, products in grouped_products:
+            products = list(sorted(products, key=attrgetter('id')))
+            # this product will only left in DB
+            main_product = products[0]
+            for product in products:
+                option = option_model(
+                    product=main_product,
+                    **{f: getattr(product, f) for f in OPTION_FIELDS},
+                    **{f: getattr(product, f) for f in TAG_FIELDS}
+                )
+                option.save()
+
+    def database_backwards(self, app_label, schema_editor, from_state, _):
+        raise NotImplementedError()
+
+
+class CleanProductsOperation(Operation):
+    """Safely remove product duplicates."""
+    def state_forwards(self, app_label, state):
+        pass
+
+    @transaction.atomic()
+    def database_forwards(self, app_label, schema_editor, _, to_state):
+        product_model = to_state.apps.get_model(app_label, 'Product')
+
+        grouped_products = groupby(
+            product_model.objects.all(),
+            key=attrgetter('name')
+        )
+        for name, products in grouped_products:
+            products = list(sorted(products, key=attrgetter('id')))
+            for product in products[1:]:
+                product.delete()
+
+    def database_backwards(self, app_label, schema_editor, from_state, _):
+        raise NotImplementedError()
+
+
+class CreateTagsOperation(Operation):
+    """Move data from product columns to tag records."""
+
+    reduces_to_sql = True
+    reversible = True
+
+    def state_forwards(self, app_label, state):
+        pass
+
+    @transaction.atomic()
+    def database_forwards(self, app_label, schema_editor, _, to_state):
+        # for every map element create new TagGroup with order value
+        group_model = to_state.apps.get_model(app_label, 'TagGroup')
+        tag_model = to_state.apps.get_model(app_label, 'Tag')
+        option_model = to_state.apps.get_model(app_label, 'Option')
+
+        field_to_group = {
+            field: group_model(name=group, position=pos)
+            for pos, (field, group) in enumerate(FIELD_TO_TAG_GROUP.items())
```

Comment:
BTW It is bad practice to rely upon dict ordering, because this aspect is implementation detail that may change in the future

Reply:
@artemiy312 , from py3.7 it's an official language feature, but not an implementation detail. If this detail will change in the future, it'll be incompatible with py3.7 language api.

But ok, in this case i'll use OrderedDict just not to worry :)